Detect compositions of {Naive}DateTime.utc_now and {Naive}DateTime.truncate

frerich opened the pull request:
This PR adds a new refactoring check which encourages people to pass a time unit to `DateTime.utc_now`.

Instead of calling `DateTime.utc_now` and then truncating the result with `DateTime.truncate/2`, it's more efficient (and less code) to just pass a time unit indicating the desired precision to `DateTime.utc_now` directly.

Dito for `NaiveDateTime`.

rrrene commented on lib/credo/check/refactor/utc_now_truncate.ex:
```diff
@@ -0,0 +1,186 @@
+      check: """
+      Pass time unit to `{Naive}DateTime.utc_now` instead of composing with `{Naive}DateTime.truncate/2`.
+
+      This should be refactored:
+
+          DateTime.utc_now() |> DateTime.truncate(:second)
+
+      to look like this:
+
+          DateTime.utc_now(:second)
+
+      Dito for `NaiveDateTime`.
```
```suggestion
      For example, the code here ...

          DateTime.utc_now() |> DateTime.truncate(:second)
          NaiveDateTime.utc_now() |> NaiveDateTime.truncate(:second)

      ... can be refactored to look like this:

          DateTime.utc_now(:second)
          NaiveDateTime.utc_now(:second)
```

rrrene commented on test/credo/check/refactor/utc_now_truncate_test.exs:
```diff
@@ -0,0 +1,265 @@
+
+  @described_check Credo.Check.Refactor.UtcNowTruncate
+
+  test "triggers when applying DateTime.truncate/2 to DateTime.utc_now/0" do
```
Please us the test name format used in most other check tests in Credo.

I think you looked at the tests for `Credo.Check.Refactor.FilterCount` as an inspiration, so this will not be your fault. 

I just pushed an update to that test suite to follow the general naming scheme of test cases in Credo: 250dbdc

frerich replied:
Sure, can do. 728bf4dc033e03faf0626fd324c60d69a47fbc01 adjusts the test names accordingly.

rrrene reviewed (changes requested):
Looking really good. Just some housekeeping items :+1:

frerich commented on lib/credo/check/refactor/utc_now_truncate.ex:
```diff
@@ -0,0 +1,186 @@
+      issue_meta,
+      message:
+        "Pass time unit to `{Naive}DateTime.utc_now` instead of composing with `{Naive}DateTime.truncate/2`.",
+      trigger: trigger,
```
Ah, I see - will fix; I suppose the bit which is triggering the warning is the `truncate/2` call, so I'll use that.

frerich replied:
155b9d5f85a33872a9e5e817c50bbc7020008489 fixes this.

rrrene replied:
I probably was unclear (sorry, not a native speaker).

`:trigger` has to be the actual piece of code causing the issue, which e.g. the editor should highlight (in this case probably `{Naive}DateTime.truncate`).

frerich replied:
Sorry, maybe I'm being a little dense here: the `trigger` is a plain string, right? I looked at other checks and gathered that it should be a name of a function causing the check to trigger.

In the check at hand, the `issue_for/3` function is invoked like

```elixir
    new_issue = issue_for(issue_meta, meta[:line], "NaiveDateTime")
```

..and its definition is

```elixir
  defp issue_for(issue_meta, line_no, module) do
    format_issue(
      issue_meta,
      message:
        "Pass time unit to `#{module}.utc_now` instead of composing with `#{module}.truncate/2`.",
      trigger: "#{module}.truncate/2",
      line_no: line_no
    )
  end
```

I.e. the third argument is a module name, and the `trigger` then becomes e.g. `DateTime.truncate/2` or `NaiveDateTime.truncate/2`, depending on what the caller passed. Is that not doing the right thing? 🤔

rrrene replied:
Yes, which is why I wrote

> `:trigger` has to be the actual piece of code causing the issue, which e.g. the editor should highlight

and there will be no `NaiveDateTime.truncate/2` in the code in your example. 

There will e.g. be `|> NaiveDateTime.truncate(:second)` and here we would want the `NaiveDateTime.truncate` part to be identified as the piece of code that "triggered" the issue (thus being the `:trigger`).

frerich replied:
Ah, I see - it needs to be a literal quote from the code, so the trigger should be just `"#{module}.truncate` then!

I suppose this also means that the `line_no` value should not be the line of the `|>` node but rather of the `truncate` expression itself. I see that `Credo.Check.add_column_if_missing/5` relies on this to calculate the `column` value for the issue. Will add another commit to do this which can be squashed later.

I wonder what this all means for aliases, e.g.

```elixir
alias DateTime, as: DT

DT.utc_now()
|>
DT.truncate(:second)
```

I suppose in this case, `DT.truncate` is the trigger to be mentioned in the issue. Will see if this can be made to work.

frerich replied:
Adjusted the check now such that it reports a correct trigger (cf. [aa498e1](https://github.com/rrrene/credo/pull/1074/commits/aa498e183a03750e624e0560940c9e02a02eec7c)) and that the line number reported is always the line in which the trigger appears (cf. [301397d](https://github.com/rrrene/credo/pull/1074/commits/301397d7aa15da895b0f2481814a77b25b64bb3f)). Looks better now? 🤗 

I briefly looked into supporting aliased uses of `DateTime`, but I believe this would amount to a significant rewrite of the check which is maybe worth a separate PR in the future. I think the logic might be roughly:

* First, traverse the AST and stop at the `defmodule` node.
* Once the `defmodule` node is visited, call `Credo.Code.Module.aliases/1` to get all aliases in the module and then start a recursive prewalk of the AST, passing the aliases along to the callback.
* In the recursive walk of the AST, don't pattern-match on the `DateTime` module directly. Instead, have each pattern extract the referenced module name and then test if the name is among the known aliases (or `DateTime` itself). Alas, since the list of aliases is not a compile-time list, this can not be done using a guard.

frerich commented on lib/credo/check/refactor/utc_now_truncate.ex:
```diff
@@ -0,0 +1,186 @@
+    format_issue(
+      issue_meta,
+      message:
+        "Pass time unit to `{Naive}DateTime.utc_now` instead of composing with `{Naive}DateTime.truncate/2`.",
```
True, improved this in 155b9d5f85a33872a9e5e817c50bbc7020008489.

frerich commented:
@rrrene thanks for the diligent review, I addressed your comments and pushed some additional commits (instead of squashing) to simplify review. If this looks okay, I guess it would be nice to squash the four most recent commits into 1a28a237e6545610ce56e20dd8996464c393b83a.

rrrene commented:
Perfect. I think we can merge this as is 🎉 

I will have to make a new branch, as we only introduce new checks in minor changes. So this will become part of Credo 1.8 ... I would target a release window of end of Jan, beginning of Feb :+1:

rrrene commented:
Hi again,

this is now part of Credo `1.7.4`. The next minor release will have to wait a bit, but I wanted to get this check in the hands of people before that.

It is now "opt-in", meaning people have to activate it. But we will include this as a standard check that is enabled by default in the future. :+1:

Sorry for the inconvenience, I should have had the idea earlier to include this as an "opt in" check for now and make it enabled by default later.

Thx again for your contribution! :tada:
